onnxruntime/core/providers/dnnl/subgraph/dnnl_relugrad.h

```diff
@@ -0,0 +1,258 @@
+namespace ort_dnnl {
+template <typename T>
+class DnnlReluGrad : public DnnlKernel {
+ public:
+  DnnlReluGrad(const DnnlNode& node,
+               DNNLExecutionProvider* provider,
+               const Provider_NodeAttributes& attributes,
+               const std::string attributes_prefix = "") : DnnlKernel(node, provider) {
+    ORT_UNUSED_PARAMETER(attributes);
+    ORT_UNUSED_PARAMETER(attributes_prefix);
+  }
+
+  void AddForwardDnnlKernel(std::shared_ptr<DnnlRelu<T>> relu_fwd) {
+    relu_fwd_ = relu_fwd;
+  }
+
+  void CreatePrimitives(const OrtCustomOpApi* api,
+                        OrtKernelContext* context,
+                        const std::unordered_map<dnnl::engine::kind, dnnl::engine>& dnnl_engine,
+                        std::vector<dnnl::primitive>& net,
+                        std::vector<std::unordered_map<int, dnnl::memory>>& net_args) override {
+    dnnl::engine cpu_engine;
+    dnnl::engine engine_to_use;
+    std::unordered_map<dnnl::engine::kind, dnnl::engine>::const_iterator iter = dnnl_engine.find(dnnl::engine::kind::cpu);
+    if (iter != dnnl_engine.end()) {
+      cpu_engine = (dnnl::engine)iter->second;
+      engine_to_use = cpu_engine;
+    }
+    #if 0 // TODO update relugrad for gpu
+    gpu_available_ = false;
+    dnnl::engine gpu_engine;
+    iter = dnnl_engine.find(dnnl::engine::kind::gpu);
+    if (iter != dnnl_engine.end()) {
+      gpu_engine = (dnnl::engine)(iter->second);
+      gpu_available_ = true;
+      engine_to_use = gpu_engine;
+      LOGS_DEFAULT(INFO) << "gpu engine found" << std::endl;
+    }
+    #endif
+    if (!relu_fwd_){
+      ORT_THROW("Unable to find forward pass Relu opt. "
+                "Verify AddForwardDnnlKernel was called after DnnlReluGrad was created.");
+    }
+
+    Ort::CustomOpApi ort{*api};
+    int input_index = mklnode_ptr_->input_start_index < 0 ? 0 : mklnode_ptr_->input_start_index;
+
+    TensorShape xgrad_shape;
+    TensorShape x_shape;
+    if (mklnode_ptr_->parent_nodes.empty()) {
+      // convert the onnx gradient Tensor to dnnl::memory and dnnl::memory::desc
+      const OrtValue* xgrad_input_tensor = ort.KernelContext_GetInput(context, input_index);
+      auto xgrad_tensor_info = ort.GetTensorTypeAndShape(xgrad_input_tensor);
+      auto xgrad_tensor_shape = ort.GetTensorShape(xgrad_tensor_info);
+      ort.ReleaseTensorTypeAndShapeInfo(xgrad_tensor_info);
+
+      auto xgradshape = xgrad_tensor_shape.data();
+      auto xgraddim = xgrad_tensor_shape.size();
+
+      dnnl::memory::dims xgraddims(xgraddim);
+
+      ort_source_format_ = GetSourceFormat(static_cast<int>(xgraddim));
+
+      xgrad_shape = TensorShape(xgradshape, xgraddim);
+
+      if (xgrad_shape.NumDimensions() == 0) {
+        primitive_created_status_ = Status(common::ONNXRUNTIME,
+                                           common::INVALID_ARGUMENT,
+                                           "Shape of size zero " + xgrad_shape.ToString());
+        return;
+      }
```

Comment:
Since the input tensor for `ReluGrad` should have the same shape and size as the corresponding `Relu` operator I reused the error logic from the `Relu` operator without giving it much thought.  We will look into this and get back to you.

Reply:
We tried to produce a test that would use a scalar tensor for Relu or ReluGrad.  We were unable to actually get Relu to run with a scalar tensor.

Reply:
Hmm, I'm surprised that Relu wouldn't accept a scalar tensor (it seems to be [implicitly allowed by the ONNX spec](https://github.com/onnx/onnx/blob/ff047ef41d5d91ac1aaa8ebe85ed449c8f0f3604/onnx/defs/math/defs.cc#L484)), although I guess it's not a very common case in NN implementations!

Either way, I still think the test should be on `xgrad_shape.Size()` to guard against returning `nullptr` from `GetTensorData()` below.

Reply:
After talking it over with developers that have been helping with this code.  I have changed the code to `Size()` as suggested. I have changed it in Relu and ReluGrad since the logic was the same in both locations.